packages/fast-components-react-base/src/auto-suggest/auto-suggest.tsx

```diff
@@ -0,0 +1,523 @@
+            nextFocusableItem.id === this.state.focusedItem.id
+        ) {
+            // at the end of the list, focus on input
+            this.focusOnInput();
+        }
+    };
+
+    /**
+     * Opens menu and focuses on first or last valid item
+     */
+    private focusOnMenu = (increment: number): void => {
+        const childrenAsArray: React.ReactNode[] = React.Children.toArray(
+            this.props.children
+        );
+
+        if (childrenAsArray.length === 0) {
+            return;
+        }
+
+        const startIndex: number = increment > -1 ? 0 : childrenAsArray.length - 1;
+        const endIndex: number = increment > -1 ? childrenAsArray.length - 1 : 0;
+        this.focusFirstItemInRange(startIndex, endIndex, childrenAsArray, increment);
+
+        this.toggleMenu(true);
+        this.shouldFocusMenuOnNextRender = true;
+    };
+
+    /**
+     * Gets first child in a range
+     */
+    private focusFirstItemInRange = (
+        startIndex: number,
+        endIndex: number,
+        childrenAsArray: React.ReactNode[],
+        increment: number
+    ): void => {
+        const validOption: React.ReactNode = Listbox.getFirstValidOptionInRange(
+            startIndex,
+            endIndex,
+            childrenAsArray,
+            increment
+        );
+        if (validOption !== null) {
+            this.updateFocusedItem([(validOption as React.ReactElement<any>).props]);
+        }
+    };
+
+    /**
+     * Toggles the menu
+     */
+    private toggleMenu = (desiredMenuState: boolean): void => {
+        const updatedIsMenuOpen: boolean = this.validateMenuState(desiredMenuState);
+        if (updatedIsMenuOpen !== this.state.isMenuOpen) {
+            this.setState({
+                isMenuOpen: updatedIsMenuOpen,
+            });
+            if (this.state.isMenuOpen && !updatedIsMenuOpen) {
+                this.focusOnInput();
+            }
+        }
+    };
+
+    /**
+     * Focus on the input element
+     */
+    private focusOnInput = (): void => {
+        const inputElements: HTMLCollectionOf<
+            HTMLInputElement
+        > = this.rootElement.current.getElementsByTagName("input");
+        if (inputElements.length > 0) {
+            inputElements[0].focus();
+        }
+        this.updateFocusedItem([]);
+    };
+
+    /**
+     * Determine menu state by comparing desired state to props
+     */
+    private validateMenuState = (desiredMenuState: boolean): boolean => {
+        let shouldOpenMenu: boolean = desiredMenuState;
+
+        if (React.Children.count(this.props.children) === 0) {
+            shouldOpenMenu = false;
+        }
+
+        if (this.props.isMenuOpen !== undefined) {
+            shouldOpenMenu = this.props.isMenuOpen;
+        }
+        return shouldOpenMenu;
+    };
+
+    /**
+     * Close the menu when when there are clicks outside
+     */
+    private handleWindowClick = (event: MouseEvent): void => {
+        if (
+            this.state.isMenuOpen &&
```

Comment:
are you manually keeping `this.state.isMenuOpen` in sync with `this.props.isMenuOpen`? 

This would probably be a good candidate for `getDerivedStateFromProps` - I think that would make this whole component more declarative.

Your whole `validateMenuState` could be moved to that function, and then you could just `setState` where needed and not have to validate the state prior to use.

Reply:
Maybe, but for the moment this is the pattern we've been following for select as well and there are places where we figure out what to do with focus based on validating menu state that may not translate well into the other appoach.  I'd take a bug to investigate making the change, but would rather not delay this component on that now (and it would be non-breaking, in theory).

Reply:
Looked at this a bit more. Since we don't have access to the component instance in the 'getDerivedState' call I'm not sure where/how we would do the focus management that we currently do in 'togglemenu'.